File: src/main/java/com/lessmarkup/engine/system/EngineConfigurationImpl.java
/*
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/.
 */
package com.lessmarkup.engine.system;

import com.lessmarkup.framework.helpers.LoggingHelper;
import com.lessmarkup.interfaces.system.EngineConfiguration;
import com.thoughtworks.xstream.XStream;

import javax.servlet.ServletConfig;
import java.io.*;
import java.util.HashMap;
import java.util.LinkedList;
import java.util.List;
import java.util.Map;

class CustomizationFile {
    private List<CustomizationFileEntry> entries;

    public List<CustomizationFileEntry> getEntries() {
        return entries;
    }

    public void setEntries(List<CustomizationFileEntry> entries) {
        this.entries = entries;
    }
}

class CustomizationFileEntry {
    private String key;
    private String value;

    public String getKey() {
        return key;
    }

    public void setKey(String key) {
        this.key = key;
    }

    public String getValue() {
        return value;
    }

    public void setValue(String value) {
        this.value = value;
    }
}

class EngineConfigurationImpl implements EngineConfiguration {

    private final String CONFIGURATION_DIRECTORY = "/config";
    private final String CONFIGURATION_FILE = CONFIGURATION_DIRECTORY + "/engine.xml";

    private final ServletConfig servletConfig;
    private final Map<String, String> overrides = new HashMap<>();
    private boolean overridesInitialized = false;

    public EngineConfigurationImpl(ServletConfig servletConfig) {
        this.servletConfig = servletConfig;
    }

    private XStream getXmlStream() {
        XStream xStream = new XStream();
        xStream.alias("file", CustomizationFile.class);
        xStream.alias("entry", CustomizationFileEntry.class);
        return xStream;
    }

    private void loadOverrides() {
        if (!overridesInitialized) {
            overridesInitialized = true;

            String configurationPath = servletConfig.getServletContext().getRealPath("/") + CONFIGURATION_FILE;

            File file = new File(configurationPath);
            if (file.exists()) {
                try (Reader reader = new FileReader(configurationPath)) {
                    XStream stream = getXmlStream();
                    CustomizationFile entries = (CustomizationFile) stream.fromXML(reader);
                    if (entries != null && entries.getEntries() != null) {
                        for (CustomizationFileEntry entry : entries.getEntries()) {
                            if (entry.getKey() != null && entry.getValue() != null) {
                                overrides.put(entry.getKey(), entry.getValue());
                            }
                        }
                    }
                } catch (IOException e) {
                    LoggingHelper.logException(getClass(), e);
                }
            }
        }
    }

    private void saveOverrides() {
        loadOverrides();

        CustomizationFile customizationFile = new CustomizationFile();
        List<CustomizationFileEntry> entries = new LinkedList<>();
        customizationFile.setEntries(entries);

        for (Map.Entry<String, String> source : overrides.entrySet()) {
            CustomizationFileEntry target = new CustomizationFileEntry();
            target.setKey(source.getKey());
            target.setValue(source.getValue());
            entries.add(target);
        }
        
        String rootPath = servletConfig.getServletContext().getRealPath("/");

        new File(rootPath + CONFIGURATION_DIRECTORY).mkdirs();

        String configurationPath = rootPath + CONFIGURATION_FILE;

        try (Writer writer = new FileWriter(configurationPath)) {
            XStream xStream = getXmlStream();
            xStream.toXML(customizationFile, writer);
        } catch (IOException e) {
            LoggingHelper.logException(getClass(), e);
        }
    }
    
    private String getString(String parameterName, String defaultValue) {
        String value = null;
        
        loadOverrides();
        value = overrides.get(parameterName);
        if (value == null) {
            value = servletConfig.getInitParameter(parameterName);
        }
        if (value == null || value.length() == 0) {
            return defaultValue;
        }
        return value;
    }

    private void setString(String parameterName, String value) {
        loadOverrides();
        overrides.put(parameterName, value);
        saveOverrides();
    }

    private boolean getBoolean(String parameterName, boolean defaultValue) {
        String value = getString(parameterName, null);
        if (value == null || value.length() == 0) {
            return defaultValue;
        }
        return value.equals("true") || !value.equals("false") && defaultValue;
    }

    private void setBoolean(String parameterName, boolean value) {
        setString(parameterName, Boolean.toString(value));
    }

    private int getInteger(String parameterName, int defaultValue) {
        String value = getString(parameterName, null);
        if (value == null || value.length() == 0) {
            return defaultValue;
        }
        return Integer.parseInt(value);
    }

    private void setInteger(String parameterName, int value) {
        setString(parameterName, Integer.toString(value));
    }

    @Override
    public boolean isSafeMode() {
        return getBoolean("safeMode", false);
    }

    @Override
    public void setSafeMode(boolean safeMode) {
        setBoolean("safeMode", safeMode);
    }

    @Override
    public String getFatalErrorsEmail() {
        return getString("fatalErrors", null);
    }

    @Override
    public void setFatalErrorsEmail(String email) {
        setString("fatalErrors", email);
    }

    @Override
    public boolean isSmtpConfigured() {
        String smtpServer = getSmtpServer();
        return smtpServer != null && smtpServer.length() > 0;
    }

    @Override
    public String getSmtpServer() {
        return getString("smtpServer", null);
    }

    @Override
    public void setSmtpServer(String server) {
        setString("smtpServer", server);
    }

    @Override
    public String getSmtpUsername() {
        return getString("smtpUsername", null);
    }

    @Override
    public void setSmtpUsername(String name) {
        setString("smtpUsername", name);
    }

    @Override
    public String getSmtpPassword() {
        return getString("smtpPassword", null);
    }

    @Override
    public void setSmtpPassword(String password) {
        setString("smtpPassword", password);
    }

    @Override
    public boolean isSmtpSsl() {
        return getBoolean("smtpSsl", false);
    }

    @Override
    public void setSmtpSsl(boolean ssl) {
        setBoolean("smtpSsl", ssl);
    }

    @Override
    public String getRecaptchaPublicKey() {
        return getString("recaptchaPublicKey", null);
    }

    @Override
    public void setRecaptchaPublicKey(String key) {
        setString("recaptchaPublicKey", key);
    }

    @Override
    public String getRecaptchaPrivateKey() {
        return getString("recaptchaPrivateKey", null);
    }

    @Override
    public void setRecaptchaPrivateKey(String key) {
        setString("recaptchaPrivateKey", key);
    }

    @Override
    public boolean isUseTestMail() {
        return getBoolean("useTestMail", false);
    }

    @Override
    public void setUseTestMail(boolean use) {
        setBoolean("useTestMail", use);
    }

    @Override
    public String getNoReplyEmail() {
        return getString("noReplyEmail", "no@reply.email");
    }

    @Override
    public void setNoReplyEmail(String email) {
        setString("noReplyEmail", email);
    }

    @Override
    public String getNoReplyName() {
        return getString("noReplyName", "NoReply");
    }

    @Override
    public void setNoReplyName(String name) {
        setString("noReplyName", name);
    }

    @Override
    public int getFailedAttemptsRememberMinutes() {
        return getInteger("failedAttemptsRememberMinutes", 15);
    }

    @Override
    public void setFailedAttemptsRememberMinutes(int minutes) {
        setInteger("failedAttemptsRememberMinutes", minutes);
    }

    @Override
    public int getMaximumFailedAttempts() {
        return getInteger("maximumFailedAttempts", 5);
    }

    @Override
    public void setMaximumFailedAttempts(int attempts) {
        setInteger("maximumFailedAttempts", attempts);
    }

    @Override
    public int getRecordsPerPage() {
        return getInteger("recordsPerPage", 10);
    }

    @Override
    public void setRecordsPerPage(int recordsPerPage) {
        setInteger("recordsPerPage", recordsPerPage);
    }

    @Override
    public String getAuthCookieName() {
        return getString("authCookieName", "LessMarkup_Auth");
    }

    @Override
    public void setAuthCookieName(String name) {
        setString("authCookieName", name);
    }

    @Override
    public int getAuthCookieTimeout() {
        return getInteger("authCookieTimeout", 15);
    }

    @Override
    public void setAuthCookieTimeout(int timeout) {
        setInteger("authCookieTimeout", timeout);
    }

    @Override
    public String getAuthCookiePath() {
        return getString("authCookiePath", "/");
    }

    @Override
    public void setAuthCookiePath(String path) {
        setString("authCookiePath", path);
    }

    @Override
    public boolean isAutoRefresh() {
        return getBoolean("autoRefresh", true);
    }

    @Override
    public void setAutoRefresh(boolean autoRefresh) {
        setBoolean("autoRefresh", autoRefresh);
    }

    @Override
    public String getNoAdminName() {
        return getString("noAdminName", "noadmin@noadmin.com");
    }

    @Override
    public void setNoAdminName(String noAdminName) {
        setString("noAdminName", noAdminName);
    }

    @Override
    public int getBackgroundJobInterval() {
        return getInteger("backgroundJobInterval", 10);
    }

    @Override
    public void setBackgroundJobInterval(int interval) {
        setInteger("backgroundJobInterval", interval);
    }

    @Override
    public String getAdminLoginPage() {
        return getString("adminLoginPage", "Login");
    }

    @Override
    public void setAdminLoginPage(String adminLoginPage) {
        setString("adminLoginPage", adminLoginPage);
    }

    @Override
    public String getAdminLoginAddress() {
        return getString("adminLoginAddress", null);
    }

    @Override
    public void setAdminLoginAddress(String adminLoginAddress) {
        setString("adminLoginAddress", adminLoginAddress);
    }

    @Override
    public boolean isMigrateDataLossAllowed() {
        return getBoolean("migrateDataLossAllowed", false);
    }

    @Override
    public void setMigrateDataLossAllowed(boolean migrateDataLossAllowed) {
        setBoolean("migrateDataLossAllowed", migrateDataLossAllowed);
    }

    @Override
    public boolean isCustomizationsDisabled() {
        return getBoolean("customizationsDisabled", false);
    }

    @Override
    public void setCustomizationsDisabled(boolean customizationsDisabled) {
        setBoolean("customizationsDisabled", customizationsDisabled);
    }

    @Override
    public String getConnectionString() {
        return getString("connectionString", null);
    }

    @Override
    public void setConnectionString(String connectionString) {
        setString("connectionString", connectionString);
    }

    @Override
    public String getModulesPath() {
        return getString("modulesPath", "modules");
    }

    @Override
    public void setModulesPath(String modulesPath) {
        setString("modulesPath", modulesPath);
    }

    @Override
    public String getSessionKey() {
        return getString("sessionKey", null);
    }

    @Override
    public void setSessionKey(String sessionKey) {
        setString("sessionKey", sessionKey);
    }
}

File: src/main/scala/com/lessmarkup/engine/system/EngineConfigurationImpl.scala
/*
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/.
 */

package com.lessmarkup.engine.system

import java.io._
import javax.servlet.ServletConfig
import com.lessmarkup.interfaces.system.EngineConfiguration
import com.thoughtworks.xstream.XStream
import scala.collection.JavaConversions._
import scala.collection.mutable
import resource._

class CustomizationFile {
  private var entries: java.util.List[CustomizationFileEntry] = null
  def getEntries: java.util.List[CustomizationFileEntry] = entries
  def setEntries(entries: java.util.List[CustomizationFileEntry]): Unit = this.entries = entries
}

class CustomizationFileEntry {
  private var key: String = null
  private var value: String = null

  def getKey = key
  def setKey(key: String): Unit = this.key = key

  def getValue: String = value
  def setValue(value: String): Unit = this.value = value
}

class EngineConfigurationImpl(servletConfig: ServletConfig) extends EngineConfiguration {

  private final val CONFIGURATION_DIRECTORY: String = ".config"
  private final val CONFIGURATION_FILE: String = CONFIGURATION_DIRECTORY + "/engine.xml"
  private lazy val overrides = loadOverrides

  def isSafeMode: Boolean = {
    getBoolean("safeMode", defaultValue = false)
  }

  def setSafeMode(safeMode: Boolean) {
    setBoolean("safeMode", safeMode)
  }

  def getFatalErrorsEmail: String = {
    getString("fatalErrors", null)
  }

  def setFatalErrorsEmail(email: String) {
    setString("fatalErrors", email)
  }

  def isSmtpConfigured: Boolean = {
    val smtpServer: String = getSmtpServer
    smtpServer != null && smtpServer.length > 0
  }

  def getSmtpServer: String = {
    getString("smtpServer", null)
  }

  def setSmtpServer(server: String) {
    setString("smtpServer", server)
  }

  def getSmtpUsername: String = {
    getString("smtpUsername", null)
  }

  def setSmtpUsername(name: String) {
    setString("smtpUsername", name)
  }

  def getSmtpPassword: String = {
    getString("smtpPassword", null)
  }

  def setSmtpPassword(password: String) {
    setString("smtpPassword", password)
  }

  protected def setString(parameterName: String, value: String) {
    overrides.put(parameterName, value)
    saveOverrides()
  }

  protected def saveOverrides() {

    val configurationPath = getConfigurationPath(createDirectory = true)
    if (configurationPath.length == 0) {
      return
    }

    val customizationFile: CustomizationFile = new CustomizationFile
    val entries: java.util.List[CustomizationFileEntry] = new java.util.LinkedList[CustomizationFileEntry]
    customizationFile.setEntries(entries)
    for (source <- overrides.entrySet) {
      val target: CustomizationFileEntry = new CustomizationFileEntry
      target.setKey(source.getKey)
      target.setValue(source.getValue)
      entries.add(target)
    }

    for {writer <- managed(new FileWriter(configurationPath))} {
      val xStream: XStream = getXmlStream
      xStream.toXML(customizationFile, writer)
    }
  }

  private def getXmlStream: XStream = {
    val xStream: XStream = new XStream
    xStream.alias("file", classOf[CustomizationFile])
    xStream.alias("entry", classOf[CustomizationFileEntry])
    xStream.alias("entries", classOf[java.util.LinkedList[CustomizationFileEntry]])
    xStream
  }

  protected def getConfigurationPath(createDirectory: Boolean = false) = {
    var rootPath: String = servletConfig.getServletContext.getRealPath("/")
    if (createDirectory) {
      new File(rootPath + CONFIGURATION_DIRECTORY).mkdirs
    }
    if (rootPath.endsWith("/") || rootPath.endsWith("\\")) {
      rootPath = rootPath.substring(0, rootPath.length - 1)
    }
    rootPath + CONFIGURATION_FILE
  }

  def isSmtpSsl: Boolean = {
    getBoolean("smtpSsl", defaultValue = false)
  }

  def setSmtpSsl(ssl: Boolean) {
    setBoolean("smtpSsl", ssl)
  }

  private def setBoolean(parameterName: String, value: Boolean) {
    setString(parameterName, value.toString)
  }

  def getRecaptchaPublicKey: String = {
    getString("recaptchaPublicKey", null)
  }

  def setRecaptchaPublicKey(key: String) {
    setString("recaptchaPublicKey", key)
  }

  def getRecaptchaPrivateKey: String = {
    getString("recaptchaPrivateKey", null)
  }

  def setRecaptchaPrivateKey(key: String) {
    setString("recaptchaPrivateKey", key)
  }

  def isUseTestMail: Boolean = {
    getBoolean("useTestMail", defaultValue = false)
  }

  private def getBoolean(parameterName: String, defaultValue: Boolean): Boolean = {
    val value: String = getString(parameterName, null)
    if (value == null || value.length == 0) {
      return defaultValue
    }
    (value == "true") || !(value == "false") && defaultValue
  }

  def setUseTestMail(use: Boolean) {
    setBoolean("useTestMail", use)
  }

  def getNoReplyEmail: String = {
    getString("noReplyEmail", "no@reply.email")
  }

  def setNoReplyEmail(email: String) {
    setString("noReplyEmail", email)
  }

  def getNoReplyName: String = {
    getString("noReplyName", "NoReply")
  }

  def setNoReplyName(name: String) {
    setString("noReplyName", name)
  }

  def getFailedAttemptsRememberMinutes: Int = {
    getInteger("failedAttemptsRememberMinutes", 15)
  }

  def setFailedAttemptsRememberMinutes(minutes: Int) {
    setInteger("failedAttemptsRememberMinutes", minutes)
  }

  def getMaximumFailedAttempts: Int = {
    getInteger("maximumFailedAttempts", 5)
  }

  private def getInteger(parameterName: String, defaultValue: Int): Int = {
    val value: String = getString(parameterName, null)
    if (value == null || value.length == 0) {
      return defaultValue
    }
    value.toInt
  }

  protected def getString(parameterName: String, defaultValue: String): String = {

    val value = overrides.get(parameterName)
    if (value.isEmpty) {
      servletConfig.getInitParameter(parameterName)
    } else if (value.get.length == 0) {
      defaultValue
    } else {
      value.get
    }
  }

  def setMaximumFailedAttempts(attempts: Int) {
    setInteger("maximumFailedAttempts", attempts)
  }

  def getRecordsPerPage: Int = {
    getInteger("recordsPerPage", 10)
  }

  def setRecordsPerPage(recordsPerPage: Int) {
    setInteger("recordsPerPage", recordsPerPage)
  }

  private def setInteger(parameterName: String, value: Int) {
    setString(parameterName, Integer.toString(value))
  }

  def getAuthCookieName: String = {
    getString("authCookieName", "LessMarkup_Auth")
  }

  def setAuthCookieName(name: String) {
    setString("authCookieName", name)
  }

  def getAuthCookieTimeout: Int = {
    getInteger("authCookieTimeout", 15)
  }

  def setAuthCookieTimeout(timeout: Int) {
    setInteger("authCookieTimeout", timeout)
  }

  def getAuthCookiePath: String = {
    getString("authCookiePath", "/")
  }

  def setAuthCookiePath(path: String) {
    setString("authCookiePath", path)
  }

  def isAutoRefresh: Boolean = {
    getBoolean("autoRefresh", defaultValue = true)
  }

  def setAutoRefresh(autoRefresh: Boolean) {
    setBoolean("autoRefresh", autoRefresh)
  }

  def getNoAdminName: String = {
    getString("noAdminName", "noadmin@noadmin.com")
  }

  def setNoAdminName(noAdminName: String) {
    setString("noAdminName", noAdminName)
  }

  def getBackgroundJobInterval: Int = {
    getInteger("backgroundJobInterval", 10)
  }

  def setBackgroundJobInterval(interval: Int) {
    setInteger("backgroundJobInterval", interval)
  }

  def getAdminLoginPage: String = {
    getString("adminLoginPage", "Login")
  }

  def setAdminLoginPage(adminLoginPage: String) {
    setString("adminLoginPage", adminLoginPage)
  }

  def getAdminLoginAddress: String = {
    getString("adminLoginAddress", null)
  }

  def setAdminLoginAddress(adminLoginAddress: String) {
    setString("adminLoginAddress", adminLoginAddress)
  }

  def isMigrateDataLossAllowed: Boolean = {
    getBoolean("migrateDataLossAllowed", defaultValue = false)
  }

  def setMigrateDataLossAllowed(migrateDataLossAllowed: Boolean) {
    setBoolean("migrateDataLossAllowed", migrateDataLossAllowed)
  }

  def isCustomizationsDisabled: Boolean = {
    getBoolean("customizationsDisabled", defaultValue = false)
  }

  def setCustomizationsDisabled(customizationsDisabled: Boolean) {
    setBoolean("customizationsDisabled", customizationsDisabled)
  }

  def getConnectionString: String = {
    getString("connectionString", null)
  }

  def setConnectionString(connectionString: String) {
    setString("connectionString", connectionString)
  }

  def getModulesPath: String = {
    getString("modulesPath", "modules")
  }

  def setModulesPath(modulesPath: String) {
    setString("modulesPath", modulesPath)
  }

  def getSessionKey: String = {
    getString("sessionKey", null)
  }

  def setSessionKey(sessionKey: String) {
    setString("sessionKey", sessionKey)
  }

  protected def loadOverrides: mutable.Map[String, String] = {

    val configurationPath = getConfigurationPath()
    val map = new mutable.HashMap[String, String]()

    if (configurationPath.length == 0) {
      return map
    }

    val file: File = new File(configurationPath)

    if (!file.exists()) {
      return map
    }

    for {reader <- managed(new FileReader(configurationPath))} {
      val stream: XStream = getXmlStream
      val entries: CustomizationFile = stream.fromXML(reader).asInstanceOf[CustomizationFile]
      if (entries != null && entries.getEntries != null) {
        for (entry <- entries.getEntries) {
          if (entry.getKey != null && entry.getValue != null) {
            map.put(entry.getKey, entry.getValue)
          }
        }
      }
    }

    map
  }
}